code/modules/atmospherics/machinery/components/binary_devices/thermomachine.dm

```diff
@@ -117,31 +121,37 @@
 		investigate_log("was set to [target_temperature] K by [key_name(user)]", INVESTIGATE_ATMOS)
 		to_chat(user, "<span class='notice'>You maximize the target temperature on [src] to [target_temperature] K.</span>")
 
-/obj/machinery/atmospherics/components/unary/thermomachine/process_atmos()
-	..()
+/obj/machinery/atmospherics/components/binary/thermomachine/process_atmos()
 	if(!is_operational || !on || !nodes[1])  //if it has no power or its switched off, dont process atmos
 		return
 	else if(is_operational && was_on == TRUE)  //if it was switched on before it turned off due to no power, turn the machine back on
 		on = TRUE
-	var/datum/gas_mixture/air_contents = airs[1]
-
-	var/air_heat_capacity = air_contents.heat_capacity()
-	var/combined_heat_capacity = heat_capacity + air_heat_capacity
-	var/old_temperature = air_contents.temperature
-
-	if(combined_heat_capacity > 0)
-		var/combined_energy = heat_capacity * target_temperature + air_heat_capacity * air_contents.temperature
-		air_contents.temperature = combined_energy/combined_heat_capacity
-
-	var/temperature_delta = abs(old_temperature - air_contents.temperature)
+	var/datum/gas_mixture/main_port = airs[1]
+	var/datum/gas_mixture/thermal_exchange_port = airs[2]
+	var/turf/local_turf = get_turf(src)
+	var/main_heat_capaticy = main_port.heat_capacity()
+	var/thermal_heat_capacity = thermal_exchange_port.heat_capacity()
+	var/temperature_delta = main_port.temperature - target_temperature
+
+	var/heat_amount = temperature_delta * (main_heat_capaticy * heat_capacity / (main_heat_capaticy + heat_capacity))
+	if(main_port.total_moles() && thermal_exchange_port.total_moles())
+		main_port.temperature = max(main_port.temperature - heat_amount / main_heat_capaticy, TCMB)
+		thermal_exchange_port.temperature = max(thermal_exchange_port.temperature + heat_amount / thermal_heat_capacity, TCMB)
```

Comment:
Add just a little bit of inefficiency and create some additional heat from the machine itself running